### ISSUE TITLE
Error events when a worker is blocked via CSP

### ISSUE THREAD

**Issue by jgraham** [web-platform-tests/interop#855]
### Description

When a worker is blocked by CSP some implementations throw a `SecurityError` however after https://github.com/whatwg/html/pull/9668 this is no longer allowed per spec.

This is an issue with observed webcompat breakage.

### Specification

https://html.spec.whatwg.org

### Additional Signals

https://bugzilla.mozilla.org/show_bug.cgi?id=1919592 
https://issues.chromium.org/issues/41285169
https://github.com/web-platform-tests/wpt/issues/41745#issuecomment-2377512723

**Comment by gsnedders** [comment]
Am I right in thinking this is all the Firefox failures in https://wpt.fyi/results/content-security-policy/worker-src?run_id=5141001291431936&run_id=6263739599028224&run_id=5140230613237760 (where the tests no longer match the spec?)

**Comment by asutherland** [comment]
> Am I right in thinking this is all the Firefox failures in https://wpt.fyi/results/content-security-policy/worker-src?run_id=5141001291431936&run_id=6263739599028224&run_id=5140230613237760 (where the tests no longer match the spec?)

Yes, with the exception of `Same-origin service worker blocked by 'none'."` and `Same-origin service worker allowed by child-src 'self'.` where Firefox is returning the wrong exception type (and which I'll fix), the problem is the the test helper [is requiring a sync throw and has a TODO since at least 2017](https://searchfox.org/mozilla-central/rev/dfc3050503739883ce6d17e8365bc35410106aba/testing/web-platform/tests/content-security-policy/support/testharness-helper.js#100-104):
```js
// TODO(mkwst): We shouldn't be throwing here. We should be firing an
// `error` event on the Worker. https://crbug.com/663298
assert_throws_dom("SecurityError", function () {
  var w = new Worker(url);
});
```

**Comment by LiangTheDev** [comment]
It looks to me that Firefox's current behavior is correct. Safari seems to have positive signal on changing its behavior at https://github.com/WebKit/standards-positions/issues/451.
I am working on changing Chromium's behavior and updating the WPT tests(tracked by https://issues.chromium.org/issues/41285169).

Do we all feel the same way that the current spec is correct and implementations should be aligned to the spec if not already?

**Comment by asutherland** [comment]
For Firefox/Gecko I think @annevk is correct that the spec is really the only sane option and we would like that to be the reality.  The synchronous throwing creates a number of implementation constraints that are undesirable, especially as it relates to nested workers. (cc @edenchuang @smaug---- just because there are more than a few threads here and they are also involved for Firefox.)

**Comment by annevk** [comment]
cc @youennf

**Comment by youennf** [comment]
> Do we all feel the same way that the current spec is correct and implementations should be aligned to the spec if not already?

The main issue is web compat. If there is too much breakage, we might have to change our mind.
We'll find out by trying to fix it.

@asutherland, do you know about that breakage in Firefox?
If the list of affected websites is not too long, we could quirk (and this list might be good for testing in any case).

**Comment by LiangTheDev** [comment]
Web compat is also the main concern for my plan to change Chromium behavior. @asutherland, do you know about site breakages in Firefox due to this difference in behavior?

**Comment by wpt-interop** [comment]
Thank you for proposing Error events when a worker is blocked via CSP for inclusion in Interop 2025.

We are pleased to let you know that this proposal was accepted as part of the Web Compat focus area. You can follow the progress of this focus area on the [Interop 2025 dashboard](https://wpt.fyi/interop-2025).

For an overview of our process, see [proposal selection](https://github.com/web-platform-tests/interop/blob/main/2025/selection-process.md). Thank you for contributing to Interop 2025!

_Posted on behalf of the Interop team._

**Comment by asutherland** [comment]
> [@asutherland](https://github.com/asutherland) could you confirm that https://bugzilla.mozilla.org/show_bug.cgi?id=1912873 can be closed now and there is currently no other known site broken due to this. Just want to avoid sites breaking immediately after we change the behavior.

I am no longer able to reproduce the problem locally on Firefox Nightly for Desktop (Linux) or Firefox Nightly for Android, so I believe Apple may have addressed its breakage.  I have set a needinfo on the bug for confirmation by the bug reporter.

We're not aware of any other current or recent site breakages attributable to complying with the spec.
